Use to AU reference table

wokenny13 opened the pull request:
This PR focuses on assigning ATTAINS Use Name to each AU in a user's WQP/TADA dataframe.

There are two functions in this PR: TADA_CreateUseAURef and TADA_CreateWaterUseRef

TADA_CreateWaterUseRef is a helper function to help assign ATTAINS Use to new/updated AUs.

This function excludes other Mod 3 functions for the time being.

See examples in the function documentations.

wokenny13 committed:
Incorporate Mod 3 spatial ref code back

wokenny13 committed:
Reflects Most Recent Updates from TADA_CreateParamRef and TADA_CreateUseParamRef PR

wokenny13 committed:
Update CriteriaMethods.R

Excel conditional formatting update

wokenny13 committed:
Update CriteriaMethods.R

handle empty column type for joining data format

wokenny13 committed:
Update ATTAINSRefTables.R

Manual additional of additional alias between ATTAINS ParameterName and WQP CharacteristicName

wokenny13 committed:
Update ATTAINSCrosswalks.R

Changed to "Not Applicable for Analysis" as the drop down value for TADA_CreateParamRef

wokenny13 committed:
Update ATTAINSCrosswalks.R

Some code styling changes and arranged drop down by ascending.

Do we want the "Not applicable for Analysis" option to show up in the top or bottom in Excel?

We mentioned top but it seems like clicking on the drop down will show bottom entries first

wokenny13 committed:
Merge branch 'develop' into Create-Module-3-AURef

wokenny13 committed:
Update GeospatialFunctions.R

Changed offset to 100 for ATTAINS geospatial webservices

wokenny13 committed:
CriteriaSummary and CriteriaTable Drafts

wokenny13 committed:
Spatial Ref table AU Bug Fix

Renamed to consistent column names between helper function and spatial crosswalk with AU spatial summary.

wokenny13 committed:
Update CriteriaMethods.R

Added Char, spec, frac to CriteriaMethods.

Reformat excel file

wokenny13 committed:
CriteriaMethod

Changed column name to ATTAINS.WaterType throughout crosswalk tables.

Edited format of excel spreadsheet outputs.

wokenny13 committed:
Update CriteriaMethodology

Edit to required columns.

DepthProfile fix incorporated

wokenny13 committed:
Mod 3 ref table formatting

cristinamullin committed:
Update DepthProfile.R

wokenny13 committed:
Merge branch 'develop' into Create-Module-3-AURef

wokenny13 committed:
Moved two spatial functions from Mod 3 to Mod 2 and Reconsideration of workflow

Reordered workflow of functions and changed function name

New order:
TADA_CreateWaterUseRef -> TADA_CreateUseAURef

Users should run these functions as part of Module 2 now.

These functions aim to help with assigning Uses to AU once sites have been joined to AUs.

wokenny13 committed:
Update CriteriaMethods.R

wokenny13 committed:
Update CriteriaMethods.R

wokenny13 committed:
Update CriteriaMethods.R

wokenny13 committed:
Update CriteriaMethods.R

wokenny13 committed:
Auto Generate Criteria Methods Option

wokenny13 committed:
Use to AU Ref Updates

wokenny13 committed:
Join Use to AUs

This PR focuses on assigning ATTAINS Use Name to each AU.

There are two functions in this PR: TADA_CreateUseAURef and TADA_CreateWaterUseRef

TADA_CreateWaterUseRef is a helper function to help assign ATTAINS Use to new/updated AUs.

This function excludes other Mod 3 functions for the time being.

wokenny13 commented on R/Tables.R:
```diff
@@ -72,7 +72,8 @@ TADA_SummarizeColumn <- function(.data, col = "TADA.CharacteristicName") {
 #' Data_6Tribes_5y_Harmonized_stats <- TADA_Stats(Data_6Tribes_5y_Harmonized)
 #'
-TADA_Stats <- function(.data, group_cols = c("TADA.ComparableDataIdentifier")) {
+TADA_Stats <- function(.data, group_cols = c("TADA.ComparableDataIdentifier"),
```
Note: remove criteriaMethods = NULL

hillarymarler committed:
Merge branch 'develop' into Use-to-AU-Reference-Table

hillarymarler commented:
I will spend some time reviewing this today, @wokenny13. Do you want review on just the functions? Or help with any check issues as well?

wokenny13 commented:
Just reviewing is fine. I can look into the check issues. Thank you!

wokenny13 commented:
• Use to AU PR
		○ User can assign **existing** uses to **existing** AUS as the default option
		○ User can assign a **user supplied list** (customized list) of uses to **existing** AUS (requires a useAURef input)
			§ The user supplied list can be **New Uses**
			§ The user supplied list can be **Existing Uses** already in ATTAINS
			§ Should we allow flexibility? 
				□ Include only the Uses in their **supplied list**
				□ Include additional rows for **both NEW AND Existing Uses** (a review process would be needed)
		○ They can assign **existing** uses to **NEW** AUs through the water type to use crosswalk (requires waterUseRef input - this can be auto generated from TADA_CreateWaterUseRef)
			§ The water type to use can be auto generated (**Existing Uses** to **Existing** Water Types)
			§ The water type to use can be user customized (**New/Existing Uses** to **Existing** Water Types)
			§ Water types should be **existing** only (found in the ATTAINS domain)
		○ They can assign a **user supplied list** of uses to **NEW** AUs through custom water type to use crosswalk (requires waterUseRef input which can be a custom table)

wokenny13 committed:
function input variable name changed

Used a sample data input for testinf and forgot to change back to sitesAURef

wokenny13 committed:
Fix Check Issues and Other Maintainence

Added global variables to utilities

Fixed Function documentation in CreateWaterUseRef

Ran styler and word list maintainence

wokenny13 committed:
Run document

Run document.

Look at the new function example for inputting a custom useAURef. Do we want this as the output where the 5 new AUs are assigned to a no use name but show as excluded?

The goal would be to identify any combinations that are not captured in the user supply useAURef table.

wokenny13 committed:
Exclude documents

Exclude CriteriaMethods.R from this PR. Not meant to be a part of this PR and will be a separate item.

hillarymarler committed:
Update ATTAINSCrosswalks.R

Add ATTAINS.watertype to output from TADA_GetATTAINSAUSiteCrosswalk

hillarymarler committed:
Merge branch 'Use-to-AU-Reference-Table' of https://github.com/USEPA/EPATADA into Use-to-AU-Reference-Table

hillarymarler committed:
Update ATTAINSCrosswalks.R

wokenny13 committed:
documentation updates

Removed missing function link, update global variables, added a missing param arg input.

hillarymarler committed:
Option for streamlining TADA_GetATTAINSParamUseOrgRef

relies on using the national download rather than looping through a web request to each org. Currently, it uses only the latest assessment cycle (similar to the old version of this function). If you wanted it to include params from all cycles (in ATTAINS) the "latest.assessments" section could be removed and the latest.params df could be generated from the original nat.assessments

hillarymarler committed:
Merge branch 'Use-to-AU-Reference-Table' of https://github.com/USEPA/EPATADA into Use-to-AU-Reference-Table

hillarymarler committed:
bug fix to retain water type

wokenny13 committed:
Removed unused function param

in TADA_CreateWaterUseRef

cristinamullin committed:
Merge branch 'develop' into Use-to-AU-Reference-Table

wokenny13 committed:
Minor text fix

Removed comment

cristinamullin commented on R/ATTAINSCrosswalks.R:
```diff
@@ -1723,3 +1760,541 @@ TADA_CreateUseParamRef <- function(.data, org_id = NULL, paramRef = NULL, usePar
+#'
+#' @export
+#'
+#' @examples
```
Suggest that this workflow using TADA_UpdateMonitoringLocationsInATTAINS with TADA_CreateUseAURef could be covered in the Module 2 vignette instead. Then we can simplify these examples to focus on TADA_CreateUseAURef.

This example makes it look like TADA_UpdateMonitoringLocationsInATTAINS is required before running TADA_CreateUseAURef. Also, the documentation mentions that the output from TADA_GetATTAINS(.data, return_sf = FALSE) can be used directly as the sitesAURef argument input in this function. Should the documentation mention TADA_UpdateMonitoringLocationsInATTAINS as well? Maybe all of that should be included in the module 2 vignette instead.

@hillarymarler @wokenny13 is this something either of you have started to work on already in a separate branch? If so, maybe we can create an issue for this and put the conversation on hold for now.

cristinamullin replied:
I ran into a few warnings when running the example code: 

Warning messages:                                                                        
1: One or more parsing issues, call `problems()` on your data frame for details, e.g.:
  dat <- vroom(...)
  problems(dat) 
2: One or more parsing issues, call `problems()` on your data frame for details, e.g.:
  dat <- vroom(...)
  problems(dat) 

```r
> # Pull a sample WQP data query
> TADA_AK_Example <- TADA_DataRetrieval(
+   startDate = "2022-01-01", endDate = "2022-12-31",
+   organization = "AKDECWQ", statecode = "AK", # update this line to reflect your spatial area of interest - whether it's by countyCode, stateCode, organizationName etc.
+   characteristicName = c("Enterococcus", "Escherichia", "Escherichia coli", "Fecal Coliform", "Total Coliform"),
+   ask = FALSE
+ )
Checking what data is available. This may take a moment.
[1] "Downloading WQP query results. This may take some time depending upon the query size."
$statecode
[1] "US:02"

$startDate
[1] "2022-01-01"

$characteristicName
[1] "Enterococcus"     "Escherichia"      "Escherichia coli" "Fecal Coliform"  
[5] "Total Coliform"  

$organization
[1] "AKDECWQ"

$endDate
[1] "2022-12-31"

[1] "Data successfully downloaded. Running TADA_AutoClean function."                     
[1] "TADA_Autoclean: creating TADA-specific columns."
[1] "TADA_Autoclean: handling special characters and coverting TADA.ResultMeasureValue and TADA.DetectionQuantitationLimitMeasure.MeasureValue value fields to numeric."
[1] "TADA_Autoclean: converting TADA.LatitudeMeasure and TADA.LongitudeMeasure fields to numeric."
[1] "TADA_Autoclean: harmonizing synonymous unit names (m and meters) to m."
[1] "TADA_Autoclean: updating deprecated (i.e. retired) characteristic names."
[1] "No deprecated characteristic names found in dataset."
[1] "TADA_Autoclean: harmonizing result and depth units."
[1] "TADA_Autoclean: creating TADA.ComparableDataIdentifier field for use when generating visualizations and analyses."
[1] "NOTE: This version of the TADA package is designed to work with numeric data with media name: 'WATER'. TADA_AutoClean does not currently remove (filter) data with non-water media types. If desired, the user must make this specification on their own outside of package functions. Example: dplyr::filter(.data, TADA.ActivityMediaName == 'WATER')"
> 
> # Alaska example to updated data links with no user supplied crosswalk
> AK_adddatalinks <- TADA_UpdateMonitoringLocationsInATTAINS(
+   org_id = "AKDECWQ",
+   crosswalk = NULL,
+   attains_replace = FALSE,
+   wqp_data_links = "none"
+ )
[1] "EQ_DomainValues: For org_id the values in the code column of the function output are the allowable values for rExpert Query functions."
[1] "EQ_AUsMLs: The current query will return 857 rows."                                 
Rows: 857 Columns: 21                                                                    
── Column specification ─────────────────────────────────────────────────────────────────
Delimiter: ","
chr (14): state, organizationType, organizationId, organizationName, waterType, monit...
dbl  (5): objectId, region, reportingCycle, cycleId, waterSize
lgl  (2): useClassName, monitoringLocationDataLink

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
[1] "TADA_GetATTAINSAUSiteCrosswalk: There are 208 monitoring location identifiers associated with assessment units for AKDECWQ in ATTAINS."
> 
> # Alaska example using a user supplied crosswalk to update entries in
> # ATTAINS by appending user supplied information to ATTAINS crosswalk
> 
> # example monitoring location identifiers
> ASSESSMENT_UNIT_ID <- c(
+   "NEW:AK_M_1021211_000", "NEW:AK_M_1021008_000",
+   "NEW:AK_M_1021109_013", "NEW:AK_M_1021109_013",
+   "NEW:AK_M_1021109_013"
+ )
> 
> # example organization identifiers
> MS_ORG_ID <- c("AKDECWQ", "AKDECWQ", "AKDECWQ", "AKDECWQ", "AKDECWQ")
> 
> # example ML, these are only examples
> MS_LOCATION_ID <- c(
+   "AKDECWQ-Snag Point", "AKDECWQ-Kanakanak", "AKDECWQ-Scandinavian",
+   "AKDECWQ-Snag Point", "AKDECWQ-Kanakanak"
+ )
> 
> ATTAINS.WaterType <- c(
+   "BEACH", "BAY", "CREEK",
+   "ESTUARY", "COASTAL"
+ )
> 
> # example urls
> MONITORING_DATA_LINK_TEXT <- c(
+   "https://www.waterqualitydata.us/provider/STORET/AKDECWQ/",
+   "https://www.waterqualitydata.us/provider/STORET/AKDECWQ/",
+   "https://www.waterqualitydata.us/provider/STORET/AKDECWQ/",
+   "https://www.waterqualitydata.us/provider/STORET/AKDECWQ/",
+   "https://www.waterqualitydata.us/provider/STORET/AKDECWQ/"
+ )
> 
> # create example crosswalk data frame
> ex.user.cw <- data.frame(
+   MS_LOCATION_ID, MS_ORG_ID, ASSESSMENT_UNIT_ID,
+   MONITORING_DATA_LINK_TEXT, ATTAINS.WaterType
+ )
> 
> AK_appenduserdata <- TADA_UpdateMonitoringLocationsInATTAINS(
+   org_id = "AKDECWQ",
+   crosswalk = ex.user.cw,
+   attains_replace = FALSE,
+   wqp_data_links = "none"
+ )
[1] "EQ_DomainValues: For org_id the values in the code column of the function output are the allowable values for rExpert Query functions."
[1] "EQ_AUsMLs: The current query will return 857 rows."                                 
Rows: 857 Columns: 21                                                                    
── Column specification ─────────────────────────────────────────────────────────────────
Delimiter: ","
chr (14): state, organizationType, organizationId, organizationName, waterType, monit...
dbl  (5): objectId, region, reportingCycle, cycleId, waterSize
lgl  (2): useClassName, monitoringLocationDataLink

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
[1] "TADA_GetATTAINSAUSiteCrosswalk: There are 208 monitoring location identifiers associated with assessment units for AKDECWQ in ATTAINS."
> 
> # Rename column for TADA_CreateUseAURef format
> AK_AU_Ref <- AK_appenduserdata %>%
+   dplyr::mutate(ATTAINS.assessmentunitidentifier = ASSESSMENT_UNIT_ID) %>%
+   dplyr::mutate(ATTAINS.OrganizationIdentifier = "AKDECWQ") %>%
+   dplyr::mutate(MonitoringLocationIdentifier = MS_LOCATION_ID) %>%
+   dplyr::select(
+     ATTAINS.assessmentunitidentifier, MonitoringLocationIdentifier,
+     ATTAINS.OrganizationIdentifier, ATTAINS.WaterType
+   )
> 
> # New AUs that are not found in ATTAINS show blank ATTAINS.UseName
> AK_CreateUseAURef <- TADA_CreateUseAURef(
+   TADA_AK_Example,
+   org_id = "AKDECWQ",
+   sitesAURef = AK_AU_Ref,
+   excel = FALSE
+ )
[1] "EQ_Assessments: The current query will return 24,081 rows."                         
Warning message:                                                                         
One or more parsing issues, call `problems()` on your data frame for details, e.g.:
  dat <- vroom(...)
  problems(dat) 
> 
> # Let's use a wateruseRef now to fill in these values.
> AK_CreateUseAURef_auto_assign <- TADA_CreateUseAURef(
+   TADA_AK_Example,
+   org_id = "AKDECWQ",
+   sitesAURef = AK_AU_Ref,
+   waterUseRef = TADA_CreateWaterUseRef(TADA_AK_EXAMPLE, org_id = "AKDECWQ"),
+   excel = FALSE
+ )
[1] "EQ_Assessments: The current query will return 24,081 rows."                         
[1] "EQ_Assessments: The current query will return 24,081 rows."                         
Warning messages:                                                                        
1: One or more parsing issues, call `problems()` on your data frame for details, e.g.:
  dat <- vroom(...)
  problems(dat) 
2: One or more parsing issues, call `problems()` on your data frame for details, e.g.:
  dat <- vroom(...)
  problems(dat) 
> 
> # We can save and reuse a useAURef as desired.
> AK_CreateUseAURef2 <- TADA_CreateUseAURef(
+   TADA_AK_Example,
+   org_id = "AKDECWQ",
+   useAURef = AK_CreateUseAURef_auto_assign,
+   sitesAURef = AK_AU_Ref,
+   excel = FALSE
+ )
[1] "EQ_Assessments: The current query will return 24,081 rows."                         
Warning message:                                                                         
One or more parsing issues, call `problems()` on your data frame for details, e.g.:
  dat <- vroom(...)
  problems(dat) 
```

wokenny13 replied:
The warning message is from the rExpertQuery functions. @hillarymarler I have been meaning to ask about this - if this means there was an error in retrieving the data or if you had other thoughts with this warning message.

wokenny13 replied:
> Is this workflow using TADA_UpdateMonitoringLocationsInATTAINS with TADA_CreateUseAURef covered in the Module 2 vignette? If so, maybe we can simplify these examples to focus on simpler TADA_CreateUseAURef examples? This example makes it look like TADA_UpdateMonitoringLocationsInATTAINS is required before running TADA_CreateUseAURef.

I don't think it's currently covered. I wanted to include an example for when new AU are added and how the helper function TADA_CreateWaterUseRef would come into play. We could use the output from TADA_GetATTAINS, but if we were to manually add examples of new AUs it might be equally long (and take longer to run). 

We can exclude the example for new AUs to make this example shorter though. And a solution with consistent column naming could help make this example shorter.

cristinamullin replied:
I think all the different examples are helpful. You can keep them here for now. This afternoon, let's discuss examples to the Module 2 vignette as well.

cristinamullin commented on R/ATTAINSCrosswalks.R:
```diff
@@ -1723,3 +1760,541 @@ TADA_CreateUseParamRef <- function(.data, org_id = NULL, paramRef = NULL, usePar
+#' Assessment Unit and Use Name Crosswalk
+#'
+#' This function pulls in all prior use names associated with each
+#' organization's assessment unit from the prior ATTAINS cycle.
```
@hillarymarler is this only AU/uses from the prior ATTAINS cycle (most recent) and not any AU/uses from any past ATTAINS cycle?

cristinamullin committed:
update TADA_CreateUseAURef documentation

cristinamullin commented on R/ATTAINSCrosswalks.R:
```diff
@@ -1723,3 +1760,561 @@ TADA_CreateUseParamRef <- function(.data, org_id = NULL, paramRef = NULL, usePar
+#' function input. This helps prevent users from overwriting their progress.
+#'
+#' @param sitesAURef A required data frame input. This data frame
+#' should contain a completed crosswalk of water types found in ATTAINS associated
```
Is water type required or optional in the ATTAINS AU/WQP site crosswalk?

wokenny13 replied:
It would be a required column as TADA_CreateWaterUseRef would need this value. I don't know if needing this though would cause any more work though for if a user already has a file with just sites/AU and not water type. I would think they would have the water types of their AUs readily available, but perhaps not? We could make Water Type not required and only required if they want to use TADA_CreateWaterUseRef. Any thoughts on this?

I was thinking about TADA_GetATTAINSAUSitesCrosswalk and how including the water type here is not exactly in the format of the ATTAINS batch upload for that tab (I think there is a separate tab to submit the water type in this). But I saw Hillary add the ATTAINS WaterType column to the TADA_GetATTAINSAUSitesCrosswalk function now.

cristinamullin commented on R/ATTAINSCrosswalks.R:
```diff
@@ -1723,3 +1760,561 @@ TADA_CreateUseParamRef <- function(.data, org_id = NULL, paramRef = NULL, usePar
+#' @param sitesAURef A required data frame input. This data frame
+#' should contain a completed crosswalk of water types found in ATTAINS associated
+#' with each assessment unit. Users will need to ensure this crosswalk contains the
+#' appropriate column names in order to run the function.
```
Ideally, users can easily generate a blank template with the required format of sitesAURef. This could be a consistent format for all TADA Module 2 functions to use for the ATTAINS AU to Monitoring Location crosswalk (sitesAURef). Let's discuss this afternoon since it related to @hillarymarler edits.

cristinamullin replied:
Should TADA_GetATTAINS generate a sitesAURef in addition to... or instead of.... appending the AU info the the dataset?

wokenny13 replied:
Hillary and I discussed this briefly, about having an argument input to make it either batch upload compatible or TADA compatible with the column naming in TADA_GetATTAINSAUCrosswalk, so that a user can then proceed with using this output as sitesAURef.

However, I have been thinking about what exactly would be the best method in handling this to get to the required format of sitesAURef. We can certainly discuss this more as I know in TADA_GetATTAINS it also ends up using a slightly different naming convention than what I have, I think in most case all lower case after the ATTAINS prefix. 

Will users always end up using TADA_Update? Will any users not desire to submit this crosswalk to ATTAINS due to not wanting to share this info?

cristinamullin commented on R/ATTAINSCrosswalks.R:
```diff
@@ -1723,3 +1760,561 @@ TADA_CreateUseParamRef <- function(.data, org_id = NULL, paramRef = NULL, usePar
+#'
+#' @param useAURef An optional data frame input. If provided, this data frame
+#' should contain a completed crosswalk of use names associated with an assessment unit.
+#' Users will need to ensure this crosswalk contains the appropriate column names in
```
same comment as above: 

Ideally, users can easily generate a blank template with the required format of useAURef. This could be a consistent format for all TADA Module 2 functions to use for the ATTAINS AU to ATTAINS use (useAURef) crosswalk.

cristinamullin replied:
in the past, we have had options to generate a blank template if the function is run with no inputs:

for example: test = TADA_CreateUseAURef()

Should this generate a blank Use/AU ref?

wokenny13 replied:
I can make this as an option for this function

cristinamullin replied:
Thanks Kenny! Can you add this option to [TADA_CreateParamRef()](https://usepa.github.io/EPATADA/reference/TADA_CreateParamRef.html) and [TADA_CreateUseParamRef()](https://usepa.github.io/EPATADA/reference/TADA_CreateUseParamRef.html) as well?

cristinamullin commented on R/ATTAINSCrosswalks.R:
```diff
@@ -1723,3 +1760,561 @@ TADA_CreateUseParamRef <- function(.data, org_id = NULL, paramRef = NULL, usePar
+#' in the current assessment cycle will not be available in ATTAINS until the
+#' assessment is approved and completed. 
+#' 
+#' @param .data A TADA dataframe. The user should run all desired data cleaning,
```
How is .data used in this function? Will this function will look for WQP Monitoring 
Location Identifier's and  ATTAINS Assessment Units in .data if they are there from running TADA_GetATTAINS first? Or is sitesAURef the source of the site/AU associations? Should TADA_GetATTAINS  include a sitesAURef output to streamline this?

wokenny13 replied:
This is good to think about.

If a user uses just the output from TADA_GetATTAINS, it will only include AU to ML assignments for those found in .data

If a user proceeds to use the output from TADA_GetATTAINSAUSitesCrosswalk, it will contain a full list of any AU to ML that an org has submitted to ATTAINS prior.

For the output of TADA_CreateUseAURef - should this focus only on the AU found in .data? Currently I believe it returns all AU from the sitesAURef. 

This is where .data would come into play, by looking through all the unique sites in .data, and only returning the AU from those sites and assigning uses to those AU.

I think it would be good to think about when deciding what is best as the output - are we also trying to make this function ATTAINS batch upload compatible?

cristinamullin committed:
add details to TADA_CreateUseAURef documentation

wokenny13 committed:
UseAURef excel return option fixes and empty dataframe creation

sitesAURef arg input changed to MLAURef name change.

if arg inputs are all blank, return empty dataframe.

will only return an empty dataframe in R, not in excel. Would users find the excel export useful if they want an empty dataframe?

For the excel output, I made a separate excel file created for this function.

wokenny13 committed:
Create empty dataframe option for ParamRef and UseParamRef

Also changed MLAURef to AUMLRef

cristinamullin commented:
See new combined PR: https://github.com/USEPA/EPATADA/pull/630
